### PR TITLE
Added trigger support in ANSI - and extended it in Postgres

### LINKED ISSUE #1815: Postgresql: Found unparsable section: DROP TRIGGER IF EXISTS my_trigger ON something;
# Expected Behaviour

The code snippet is parsed successfully.

# Observed Behaviour

PRS error

`L:  41 | P:   1 |  PRS | Line 41, Position 1: Found unparsable section: DROP TRIGGER IF EXISTS my_trigger ON something;'`

# Steps to Reproduce

Parse this with postgres dialect:

`DROP TRIGGER IF EXISTS my_trigger ON something;`

# Configuration

- Command: `sqlfluff lint --dialect=postgres --exclude-rules L039,L016 --processes 2 --ignore PRS path/to/sql/dir-files`
- SQLfluff 0.7.1
- Python 3.9.x

### PR DESCRIPTION
**Pull request by WittierDinosaur**
### Brief summary of the change made
Added CreateTrigger and DropTrigger statement segments in ANSI. ANSI doesn't support Alter Trigger.
Added AlterTrigger in Postgres - and extended Create and Drop to match the PG14 docs. ANSI definitions were reverse engineered from the Postgres docs (what they said they didn't support).

Fixes #1815

Partially addresses #1487

### Are there any other side effects of this change that we should be aware of?
No

### Pull Request checklist
Docstrings with references to the docs added.
Test cases added and passing.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -3157,3 +3166,86 @@ class DatePartFunctionNameSegment(BaseSegment):
...
+        OneOf("BEFORE", "AFTER", Sequence("INSTEAD", "OF")),
+        Delimited(
+            "INSERT",
+            "DELETE",
+            Sequence(
+                "UPDATE",
+                "OF",
+                Delimited(
+                    Ref("ColumnReferenceSegment"),
+                    terminator=OneOf("INSERT", "DELETE", "UPDATE", "ON"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_ansi.py:3196]: Do "INSERT", "DELETE", "UPDATE" need to be in here?

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 2]
```diff
@@ -3157,3 +3166,86 @@ class DatePartFunctionNameSegment(BaseSegment):
...
+                ),
+            ),
+            delimiter="OR",
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_ansi.py:3199]: Not seen two character `delimiters` before and no test case for this. Could we add one just to make sure it works?

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 3]
```diff
@@ -3157,3 +3166,86 @@ class DatePartFunctionNameSegment(BaseSegment):
...
+            Sequence(
+                "REFERENCING",
+                "OLD",
+                "ROW",
+                "AS",
+                Ref("ParameterNameSegment"),
+                "NEW",
+                "ROW",
+                "AS",
+                Ref("ParameterNameSegment"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_ansi.py:3214]: The Postgres page specifically gives examples of:
- Only setting the `NEW` reference
- Setting both but the other way around.

Are these definitely not supported in ANSI?

Given this is in an `AnyNumberOf`, couldn't we just do this?:

```suggestion
                "REFERENCING",
                OneOf("OLD", "NEW"),
                "ROW",
                "AS",
                Ref("ParameterNameSegment"),
```

File: test/fixtures/dialects/postgres/postgres_create_trigger.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+CREATE TRIGGER foo AFTER UPDATE OF bar, baz ON bar EXECUTE FUNCTION proc(args);
+
+CREATE TRIGGER foo INSTEAD OF DELETE ON bar FROM baz DEFERRABLE INITIALLY DEFERRED FOR EACH STATEMENT EXECUTE FUNCTION proc(args);
+
+CREATE TRIGGER foo BEFORE INSERT ON bar WHEN (a=b) EXECUTE FUNCTION proc(args);
+
+CREATE OR REPLACE CONSTRAINT TRIGGER foo BEFORE INSERT ON bar EXECUTE FUNCTION proc(args);
+
+CREATE TRIGGER foo BEFORE INSERT ON bar REFERENCING OLD TABLE as old_table NEW TABLE AS new_table EXECUTE PROCEDURE proc(args);
```
Comment (tunetheweb) [inline test/fixtures/dialects/postgres/postgres_create_trigger.sql:13]: Could you include the examples from https://www.postgresql.org/docs/14/sql-createtrigger.html?

And similarly for the other tests.

Always good to include the examples when they're good enough to give.

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 1 above)
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:3196]: Ah yes - that should have been OR or ON

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 2 above)
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:3199]: Tested - it works!

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 3 above)
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:3214]: This specific part of syntax is not supported by Postgres - I think you're thinking of the Table Referencing.

File: test/fixtures/dialects/postgres/postgres_create_trigger.sql (hunk 4 above)
Comment (WittierDinosaur) [reply test/fixtures/dialects/postgres/postgres_create_trigger.sql:13]: Sorted - bar one. Fixed another issue getting one passing, adding another issue for the other. Can't getr `table.*` to parse

Comment (tunetheweb) [reply test/fixtures/dialects/postgres/postgres_create_trigger.sql:13]: Want to merge as is and come back to this? Or still working on it?

Comment (WittierDinosaur) [reply test/fixtures/dialects/postgres/postgres_create_trigger.sql:13]: I've added an issue - let's merge as is. I'll come back to this once 0.8 is out the door

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**